src/main/java/io/vertx/core/eventbus/impl/clustered/ClusteredEventBus.java

```diff
@@ -206,31 +193,18 @@ public MessageImpl createMessage(boolean send, String address, MultiMap headers,
   protected <T> void sendOrPub(OutboundDeliveryContext<T> sendContext) {
     if (((ClusteredMessage) sendContext.message).getRepliedTo() != null) {
       clusteredSendReply(((ClusteredMessage) sendContext.message).getRepliedTo(), sendContext);
+    } else if (sendContext.options.isLocalOnly()) {
+      super.sendOrPub(sendContext);
     } else {
-      if (sendContext.options.isLocalOnly()) {
-        super.sendOrPub(sendContext);
-      } else if (Vertx.currentContext() != sendContext.ctx) {
-        // Current event-loop might be null when sending from non vertx thread
-        sendContext.ctx.runOnContext(v -> {
-          subs.get(sendContext.message.address(), ar -> onSubsReceived(ar, sendContext));
-        });
-      } else {
-        subs.get(sendContext.message.address(), ar -> onSubsReceived(ar, sendContext));
-      }
-    }
-  }
-
-  private <T> void onSubsReceived(AsyncResult<ChoosableIterable<ClusterNodeInfo>> asyncResult, OutboundDeliveryContext<T> sendContext) {
-    if (asyncResult.succeeded()) {
-      ChoosableIterable<ClusterNodeInfo> serverIDs = asyncResult.result();
-      if (serverIDs != null && !serverIDs.isEmpty()) {
-        sendToSubs(serverIDs, sendContext);
-      } else {
-        super.sendOrPub(sendContext);
-      }
-    } else {
-      log.error("Failed to send message", asyncResult.cause());
-      sendContext.written(asyncResult.cause());
+      sendContext.ctx.succeededFuture(sendContext.message)
```

Comment:
avoid using `flatMap`, instead simply call `deliveryStrategy.chooseNodes(sendContext.message)`. The current pattern does not do much and consumes allocation.

If we need to use future for proper message delivery then instead, change the contract of `chooseNodes` to use a `Promise<>List<String>` and let the caller pass a promise that defines the context on which the callback is done.

Reply:
The current pattern allows to invoke the delivery strategy on the `sendContext.ctx`.

If we remove the creation of the `succeededFuture`, we'll still have to invoke `runOnContext` somewhere. This will imply allocation too, won't it?

Reply:
why is it necessary to call the delivery strategy on a particular context ? I feel it should not depend on a particular context as the eventual synchronisation of the delivery strategy is not related to vertx concurrency, or am I missing something here ?